numpy/typing/_callable.py

```diff
@@ -312,6 +314,12 @@ def __call__(
     class _NumberOp(Protocol):
         def __call__(self, __other: _NumberLike) -> number: ...
 
+    class _ComparisonOp(Protocol[_T]):
+        @overload
+        def __call__(self, __other: _T) -> bool_: ...
```

Comment:
Are these overloads in the right order? I can't find docs so quickly, but I had expected overloads to be checked in order, so `_T` would also match `ArrayLike`.

Reply:
Yes, in practice the first overload is used for all valid scalar-types, which always return `bool_`, while the second one is used for the much broader `ArrayLike` (a superset of the first overload).

Reply:
Also note that `_T` is a typevar, so it should be substituted with an actual value.
For example, `_ComparisonOp[datetime64]` replaces `_T` with `datetime64`.

Reply:
> Yes, in practice the first overload is used for all valid scalar-types

Thanks. That's what I was (am) fuzzy about. The Python docs on `TypeVar` are pretty bad, and PEP 484 says _"By default, a type variable ranges over all possible types."_. My interpretation was really "all", not "all scalar". You can use `TypeVar( ...bound=`  with classes, mappings, etc. as bounds. So I thought `_T` would match array-likes as well here.

Reply:
For an unbound (_i.e._ unassigned) typevars this statement is actually true, even if the `bound` parameter is specified (because reasons...). This is the very very reason why it is recommended to always assign some value to subscriptable types, _e.g._ use `Sequence[Any]` instead of plain `Sequence`.

The above-mentioned fortunately becomes irrelevant once a value has been assigned to the typevar (`Seq3` in the example below).

``` python
from typing import TypeVar, Sequence

T = TypeVar("T", bound=int)

Seq1 = Sequence  # equivalent to `Sequence[Any]`
Seq2 = Sequence[T]  # equivalent to `Sequence[Any]`
Seq3 = Seq2[int]  # equivalent to `Sequence[int]`
```

Reply:
Thanks, that makes more sense now.